### PR TITLE
Parse GATs in associated type args

### LINKED ISSUE #9602: Generic Associated Types fail to parse when used in constraints
[Rust Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=26614623eb7aaba3a370e9d670701c5b) of code which fails to parse in Rust-Analyzer; specifically, see line 24:
`Inner: for<'a> Foo<Bar<'a> = &'a str>`

GAT specifications in constraints cause the parser to abort; RA `expected COMMA` and does not appear to be usable with GAT constraint specifications in any expressible form.

The sort of constraint that fails, and what RA displays, despite the code being legal and compiling successfully:
image

### PR DESCRIPTION
**Pull request by Veykril**
Fixes #9602

### CODE REVIEW

**Comment by Dessix** [comment]
This may solve #9602.

File: Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,6 @@ debug = 0 # Set this to 1 or 2 to get more useful backtraces in debugger.
 # chalk-ir = { path = "../chalk/chalk-ir" }
 # chalk-recursive = { path = "../chalk/chalk-recursive" }
 
-# ungrammar = { path = "../ungrammar" }
+ungrammar = { path = "../ungrammar" }
```
Comment (Dessix) [inline Cargo.toml:31]: Including this in your changeset may be the cause of the current CI failures.

Comment (Veykril) [reply Cargo.toml:31]: Ye thats expected, will have to merge https://github.com/rust-analyzer/ungrammar/pull/36 first and then point to the updated version there

File: crates/parser/src/grammar/type_args.rs [hunk 2]
```diff
@@ -65,17 +65,48 @@ fn generic_arg(p: &mut Parser) {
...
             name_ref(p);
-            type_params::bounds(p);
-            m.complete(p, ASSOC_TYPE_ARG);
-        }
-        IDENT if p.nth(1) == T![=] => {
-            name_ref(p);
-            p.bump_any();
-            types::type_(p);
-            m.complete(p, ASSOC_TYPE_ARG);
+            if p.current() == T![<] {
```
Comment (Dessix) [inline crates/parser/src/grammar/type_args.rs:74]: Looks like we're already in a `match p.current()` by this point, with its value currently bound to `IDENT`; would it be more intuitive to query against `IDENT` instead?

File: crates/parser/src/grammar/type_args.rs [hunk 3]
```diff
@@ -65,17 +65,48 @@ fn generic_arg(p: &mut Parser) {
...
+            if p.current() == T![=] {
+                // NameRef<...> =
+                p.bump_any();
+                types::type_(p);
+
+                path_seg.abandon(p);
+                path.abandon(p);
+                path_ty.abandon(p);
+                m.complete(p, ASSOC_TYPE_ARG);
+            } else if p.current() == T![:] {
```
Comment (Dessix) [inline crates/parser/src/grammar/type_args.rs:86]: We've got a chain of `p.current()` comparisons against constants here. A `match IDENT` against the macro-ized constants might be more fitting.

File: crates/parser/src/grammar/type_args.rs (hunk 2 above)
Comment (Veykril) [reply crates/parser/src/grammar/type_args.rs:74]: The cursor position here has been moved due to the name_ref call prior to this.

**Pull request merged**